doc/design/drafts/dynamic_size_types.md

```diff
@@ -0,0 +1,496 @@
+   container class.
+3. A function developer should not have to specify the allocator type of a
+   container. This means the allocator type is independent of the container type.
+   This has to work:
+   ```cpp
+   void myFunkyFunction(const cxx::vector<int> & bla) {}
+
+   cxx::vector<int, 20> a;
+   cxx::vector<int, 30> b;
+   cxx::vector<int> c(Allocator::heap);
+
+   myFunkyFunction(a);
+   myFunkyFunction(b);
+   myFunkyFunction(c);
+   ```
+4. All containers must be shared memory compatible, this means no vtables and
+   virtual. Furthermore the usage of function pointers, `cxx::function` and
+   `cxx::function_ref` is forbidden in all containers and their underlying 
+   constructs.
+5. Before using any cxx container the user as to call `reserve()` once otherwise
+   the container will have a default capacity of zero. It is not allowed to
+   call `reserve()` multiple times since this could violate the zero copy
+   guarantee.
+   Stack based containers do not require a `reserve()` call beforehand but
+   should provide this call as well for compatibility reasons.
+
+The draft we provide here is using the `cxx::vector` but the techniques
+described can be easily applied to the `cxx::string` or other cxx containers.
+
+```
++------------------------------------------------------+
+|template<T>                                           |
+|vector                                                |
+|                                                      |
+|  - vector() = default                                |
+|  - template<Allocator>                               |
+|    vector(Allocator & allocator);                    |
+|                                                      |
+|  - expected<AllocatorError> reserve(const uint64_t); |
+|  - void release();                                   |
+|                                                      |
+|  // remaining API is unchanged                       |
++------------------------------------------------------+
+                          ^
+                         /|\
+                          |
+       +------------------+--------------------------+
+       |template<T, Capacity>                        |
+       |vector : public vector<Type>                 |
+       |                                             |
+       |  (#1)                                       |
+       |  - vector(const vector<T> &)                |
+       |  - vector(vector<T> &&)                     |
+       |  - vector& operator=(const vector<T> &)     |
+       |  - vector& operator=(vector<T> &&)          |
+       |                                             |
+       |                                             |
+       |  // same API as current cxx::vector         |
+       |                                             |
+       |  # uint8_t m_data[Capacity]                 |
+       |  # RangeAllocator m_allocator               |
+       +---------------------------------------------+
```

Comment:
I added this idea to the design document under open questions to keep this idea documented.

Reply:
Not sure what `payload_vector` is here, I probably missed something.

But we may also request the allocator from the publisher in a way. Then we could provide this allocator to the vector as sketched in the document, but of course we would have to be careful with the lifetime.

Of course the vector would now have a pointer to the allocator, which I think the proposal here tries to avoid.
I think 2. is very error-prone and cumbersome to use.